### PR TITLE
Redesign sequencer interface

### PR DESCRIPTION
**Pull request by apmasell**
This makes several UI changes:

- drops all the stats UI components
- upgrades the list sequencers page to the new infrastructure
- provides a new non-stats way to add a sequencer

### CODE REVIEW

File: miso-web/src/main/webapp/scripts/list_sequencer.js [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        "handler" : function() {
+          
+          Utils.showDialog("Add Sequencer", "Add", [ {
+            property : "name",
+            type : "text",
+            label : "Name"
+          }, {
+            property : "platform",
+            type : "select",
+            values : Constants.platforms,
```
Comment (djcooke) [inline miso-web/src/main/webapp/scripts/list_sequencer.js]: Sort by platform and model

File: miso-web/src/main/webapp/scripts/list_sequencer.js [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+          }, {
+            property : "serialNumber",
+            type : "text",
+            label : "Serial Number"
+          }, {
+            property : "ip",
+            type : "text",
+            label : "Hostname/IP Address"
+          }, {
+            property : "dateCommissioned",
```
Comment (djcooke) [inline miso-web/src/main/webapp/scripts/list_sequencer.js]: A date picker attached to a text field like we use on most forms would be clearer. This is an optional field and while no date is selected initially, it looks like the current date is. There doesn't seem to be a way to deselect a date after selecting one by accident either

File: miso-web/src/main/webapp/scripts/list_sequencer.js [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+  createStaticActions : function(config, projectId) {
+    if (config.isAdmin) {
+      return [ {
+        "name" : "Add",
+        "handler" : function() {
+          
+          Utils.showDialog("Add Sequencer", "Add", [ {
+            property : "name",
+            type : "text",
+            label : "Name"
```
Comment (djcooke) [inline miso-web/src/main/webapp/scripts/list_sequencer.js]: Use `*` beside labels to indicate required fields

File: miso-web/src/main/webapp/scripts/list_sequencer.js [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+            property : "ip",
+            type : "text",
+            label : "Hostname/IP Address"
+          }, {
+            property : "dateCommissioned",
+            type : "date",
+            label : "Date Comissioned"
+          } ], function(seq) {
+            seq.id = 0;
+            Utils.ajaxWithDialog('Saving Sequencer', 'POST',
```
Comment (djcooke) [inline miso-web/src/main/webapp/scripts/list_sequencer.js]: If this fails for any reason, it just shows "Sadness: error." Check required fields before submitting and set a better default error message?

File: miso-dto/src/main/java/uk/ac/bbsrc/tgac/miso/dto/Dtos.java [hunk 5]
```diff
@@ -1858,4 +1860,28 @@ public static GroupDto asDto(Group from) {
...
+  public static SequencerReference to(SequencerDto dto) {
+    SequencerReference to = new SequencerReferenceImpl();
+    to.setId(dto.getId());
+    to.setDateCommissioned(extractDateOrNull(dto.getDateCommissioned()));
+    to.setDateDecommissioned(extractDateOrNull(dto.getDateDecommissioned()));
+    to.setIpAddress(dto.getIp());
+    to.setName(dto.getName());
+    to.setSerialNumber(dto.getSerialNumber());
+    to.setPlatform(to(dto.getPlatform()));
+    to.setSerialNumber(dto.getSerialNumber());
```
Comment (djcooke) [inline miso-dto/src/main/java/uk/ac/bbsrc/tgac/miso/dto/Dtos.java]: set twice

File: miso-dto/src/main/java/uk/ac/bbsrc/tgac/miso/dto/Dtos.java [hunk 6]
```diff
@@ -1858,4 +1860,28 @@ public static GroupDto asDto(Group from) {
...
+
+  public static SequencerDto asDto(SequencerReference from) {
+    SequencerDto dto = new SequencerDto();
+    dto.setId(from.getId());
+    dto.setDateCommissioned(getDateString(from.getDateCommissioned()));
+    dto.setDateDecommissioned(getDateString(from.getDateDecommissioned()));
+    dto.setName(from.getName());
+    dto.setPlatform(asDto(from.getPlatform()));
+    dto.setSerialNumber(from.getSerialNumber());
+    return dto;
```
Comment (djcooke) [inline miso-dto/src/main/java/uk/ac/bbsrc/tgac/miso/dto/Dtos.java]: ipAddress not set

File: miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultSequencerReferenceService.java [hunk 7]
```diff
@@ -38,7 +41,7 @@
 
   @Override
   public Collection<SequencerReference> listByPlatformType(PlatformType platformType) throws IOException {
-    return sequencerReferenceDao.listByPlatformType(platformType);
+    return sequencerReferenceDao.list(0, 0, true, "id", PaginationFilter.archived(false));
```
Comment (djcooke) [inline miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultSequencerReferenceService.java]: Method name is `listByPlatformType`, but that's not what it's doing anymore? PlatformType parameter is being ignored

File: miso-web/src/main/webapp/pages/editSequencerReference.jsp [hunk 8]
```diff
@@ -36,7 +36,7 @@
 
 <div id="maincontent">
   <div id="contentcolumn">
-    <form:form id="sequencer_reference_form" data-parsley-validate="" action="/miso/stats/sequencer" method="POST" commandName="sequencerReference" autocomplete="off">
+    <form:form id="sequencer_reference_form" data-parsley-validate="" action="/miso/sequencer" method="POST" commandName="sequencerReference" autocomplete="off">
```
Comment (djcooke) [inline miso-web/src/main/webapp/pages/editSequencerReference.jsp]: Save is attempting POST to `/miso/sequencer` and 404ing

File: miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/SequencerRestController.java [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+import uk.ac.bbsrc.tgac.miso.core.util.PaginatedDataSource;
+import uk.ac.bbsrc.tgac.miso.dto.DataTablesResponseDto;
+import uk.ac.bbsrc.tgac.miso.dto.Dtos;
+import uk.ac.bbsrc.tgac.miso.dto.SequencerDto;
+import uk.ac.bbsrc.tgac.miso.service.SequencerReferenceService;
+
+@Controller
+@RequestMapping("/rest/sequencer")
+public class SequencerRestController extends RestController {
+  private final JQueryDataTableBackend<SequencerReference, SequencerDto> jQueryBackend = new JQueryDataTableBackend<SequencerReference, SequencerDto>() {
```
Comment (djcooke) [inline miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/SequencerRestController.java]: Can be static

Comment (apmasell) [inline miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/SequencerRestController.java]: It can't. It depends on an instance field.

File: miso-web/src/main/webapp/scripts/list_sequencer.js (hunk 2 above)
Comment (apmasell) [inline miso-web/src/main/webapp/scripts/list_sequencer.js]: It's not supposed to be optional, so mandatory selection was by design.

**Comment by apmasell** [comment]
`StatsControllerHelperService` and `stats_ajax.js` need to wait until #1014 is also merged.

**Comment by apmasell** [comment]
Everything else is fixed.

**Comment by djcooke** [comment]
Date commissioned is actually supposed to be an optional field. After setting it in the create dialog, you can still remove it on the Edit Sequencer page. I don't think anything depends on it. Is there a reason it should be required?

The add sequencer dialog is now huge. I guess it doesn't matter how it looks since it will be rarely used and only by admin though.

**Comment by djcooke** [comment]
I'm getting a stack trace on the pools, containers, runs, and indices list pages:

```
25-Jul-2017 11:50:47.135 SEVERE [http-nio-8080-exec-6] org.apache.catalina.core.StandardWrapperValve.invoke Servlet.service() for servlet
 [miso] in context with path [] threw exception [Request processing failed; nested exception is org.springframework.orm.hibernate4.Hibern
ateSystemException: Could not obtain transaction-synchronized Session for current thread; nested exception is org.hibernate.HibernateExce
ption: Could not obtain transaction-synchronized Session for current thread] with root cause
 org.hibernate.HibernateException: Could not obtain transaction-synchronized Session for current thread
        at org.springframework.orm.hibernate4.SpringSessionContext.currentSession(SpringSessionContext.java:134)
        at org.hibernate.internal.SessionFactoryImpl.getCurrentSession(SessionFactoryImpl.java:993)
        at uk.ac.bbsrc.tgac.miso.persistence.impl.HibernateSequencerReferenceDao.currentSession(HibernateSequencerReferenceDao.java:75)
        at uk.ac.bbsrc.tgac.miso.persistence.impl.HibernatePaginatedDataSource.createPaginationCriteria(HibernatePaginatedDataSource.java
:47)
        at uk.ac.bbsrc.tgac.miso.persistence.impl.HibernatePaginatedDataSource.list(HibernatePaginatedDataSource.java:80)
        at uk.ac.bbsrc.tgac.miso.core.util.PaginatedDataSource.list(PaginatedDataSource.java:33)
        at uk.ac.bbsrc.tgac.miso.core.util.PaginatedDataSource$$FastClassBySpringCGLIB$$f501affa.invoke(<generated>)
        at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204)
        at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:721)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157)
        at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.j
ava:136)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
        at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:656)
        at uk.ac.bbsrc.tgac.miso.persistence.impl.HibernateSequencerReferenceDao$$EnhancerBySpringCGLIB$$d68acb6b.list(<generated>)
        at uk.ac.bbsrc.tgac.miso.core.manager.MisoRequestManager.listActivePlatformTypes(MisoRequestManager.java:444)
        at uk.ac.bbsrc.tgac.miso.core.manager.UserAuthMisoRequestManager.listActivePlatformTypes(UserAuthMisoRequestManager.java:271)
        at uk.ac.bbsrc.tgac.miso.webapp.util.TabbedListItemsPage.getPlatformTypes(TabbedListItemsPage.java:44)
        at uk.ac.bbsrc.tgac.miso.webapp.util.TabbedListItemsPage.createForPlatformType(TabbedListItemsPage.java:27)
        at uk.ac.bbsrc.tgac.miso.webapp.controller.ListIndicesController.listIndices(ListIndicesController.java:33)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
```

Does not happen on current develop

**Comment by apmasell** [comment]
Fixed.

File: miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditSequencerReferenceController.java [hunk 10]
```diff
@@ -25,111 +25,91 @@
...
+  public String processSubmit(@ModelAttribute("sequencerReference") SequencerReference sr, ModelMap model, SessionStatus session)
+      throws IOException {
+    Long srId = null;
+    if (sr.getId() == SequencerReferenceImpl.UNSAVED_ID) {
+      srId = sequencerReferenceService.create(sr);
+    } else {
+      sequencerReferenceService.update(sr);
+      srId = sr.getId();
     }
+    return "redirect:/miso/sequencer/" + srId;
```
Comment (callunity) [inline miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditSequencerReferenceController.java]: Editing a sequencer saves successfully, but sets the page's URL to this: `http://localhost:8080/miso/sequencer/49?misoBugUrl=https%3A%2F%2Fjira.oicr.on.ca%2Fsecure%2FCreateIssueDetails%21init.jspa%3Fissuetype%3D1%26pid%3D11685&misoInstanceName=Heather%27s+Test+Kitchen`

Comment (apmasell) [inline miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditSequencerReferenceController.java]: Fixed.

**Comment by djcooke** [comment]
The resized dialog actually affected other dialogs. The create and propagate sample dialogs are now huge too

**Comment by apmasell** [comment]
The jQuery dialog system doesn't have a way to dynamically resize based on content.

**Comment by djcooke** [comment]
The function in lims.js that creates the dialog could accept sizing parameters

**Comment by apmasell** [comment]
How's this?
